client/scripts/mifarePlus.lua

```diff
@@ -0,0 +1,284 @@
+local cmds = require('commands')
+local lib14a = require('read14a')
+
+SIXTEEN_BYTES_ZEROS = "00000000000000000000000000000000"
+
+GETVERS_INIT = "0360" -- Begins the GetVersion command
+GETVERS_CONT = "03AF" -- Continues the GetVersion command
+POWEROFF = "OFF"
+WRITEPERSO = "03A8"
+COMMITPERSO = "03AA"
+AUTH_FIRST = "0370"
+AUTH_CONT = "0372"
+AUTH_NONFIRST = "0376"
+PREPAREPC = "03F0"
+PROXIMITYCHECK = "03F2"
+VERIFYPC = "03FD"
+READPLAINNOMACUNMACED = "0336"
+
+--- 
+-- This is only meant to be used when errors occur
+function oops(err)
+	print("ERROR: ",err)
+end
+
+---
+-- Used to send raw data to the firmware to subsequently forward the data to the card.
+function sendRaw(rawdata, crc, power)
+	print(("<sent>: 	  %s"):format(rawdata))
+
+	local flags = lib14a.ISO14A_COMMAND.ISO14A_RAW
+	if crc then
+		flags = flags + lib14a.ISO14A_COMMAND.ISO14A_APPEND_CRC
+	end
+	if power then
+		flags = flags + lib14a.ISO14A_COMMAND.ISO14A_NO_DISCONNECT
+	end
+
+	local command = Command:new{cmd = cmds.CMD_READER_ISO_14443a, 
+								arg1 = flags, -- Send raw
+								arg2 = string.len(rawdata) / 2, -- arg2 contains the length, which is half the length of the ASCII-string rawdata
+								data = rawdata}
+	local ignore_response = false
+	local result, err = lib14a.sendToDevice(command, ignore_response)
+	if result then
+		--unpack the first 4 parts of the result as longs, and the last as an extremely long string to later be cut down based on arg1, the number of bytes returned
+		local count,cmd,arg1,arg2,arg3,data = bin.unpack('LLLLH512',result)
+		returned_bytes = string.sub(data, 1, arg1 * 2)
+		print(("<recvd>: %s"):format(returned_bytes)) -- need to multiply by 2 because the hex digits are actually two bytes when they are strings
+		return returned_bytes
+	else
+		err = "Error sending the card raw data."
+		oops(err)
+	end
+end
+
+function writePerso()
+	-- Used to write any data, including the keys (Key A and Key B), for all the sectors.
+	-- writePerso() command parameters:
+	--		1 byte  - 0xA8 - Command Code
+	--		2 bytes - Address of the first block or key to be written to (40 blocks are numbered from 0x0000 to 0x00FF)
+	--		X bytes - The data bytes to be written, starting from the first block. Amount of data sent can be from 16 to 240 bytes in 16 byte increments. This allows
+	--				up to 15 blocks to be written at once.
+	-- response from PICC:
+	--		0x90 - OK
+	--		0x09 - targeted block is invalid for writes, i.e. block 0, which contains manufacturer data
+	--		0x0B - command invalid
+	--		0x0C - unexpected command length
+	
+
+	-- First, set all the data in the card (4kB of data) to zeros. The keys, stored in the sector trailer block, are also set to zeros.
```

Comment:
What about 2K and 8K cards?

Anyway: do we need this? Keys can only be written in SL0. Can't we assume that the data blocks are empty for a card in SL0?

Reply:
It would be best to explicitly set the data in the card in SL0, especially because the sector trailers have the keys which are used for Mifare Classic backwards compatibility. We could go back and set them later, but it is easier to do it in SL0. I have added in a fix for writing to 2kB and 4kB cards (not the prettiest, but we only CommitPerso once per card).